utils/signature/verify.go

```diff
@@ -7,58 +7,121 @@ SPDX-License-Identifier: Apache-2.0
+	"github.com/hyperledger/fabric-x-common/common/policies"
+	"github.com/hyperledger/fabric-x-common/msp"
+	"github.com/hyperledger/fabric-x-common/protoutil"
 
 	"github.com/hyperledger/fabric-x-committer/api/protoblocktx"
 )
 
-// DigestVerifier verifies a digest.
-type DigestVerifier interface {
-	Verify(Digest, Signature) error
-}
-
 // NsVerifier verifies a given namespace.
 type NsVerifier struct {
-	verifier  DigestVerifier
-	Scheme    Scheme
-	PublicKey PublicKey
+	verifier        policies.Policy
+	NamespacePolicy *protoblocktx.NamespacePolicy
 }
 
 // NewNsVerifier creates a new namespace verifier according to the implementation scheme.
-func NewNsVerifier(scheme Scheme, key PublicKey) (*NsVerifier, error) {
+func NewNsVerifier(p *protoblocktx.NamespacePolicy, idDeserializer msp.IdentityDeserializer) (*NsVerifier, error) {
 	res := &NsVerifier{
-		Scheme:    strings.ToUpper(scheme),
-		PublicKey: key,
+		NamespacePolicy: p,
 	}
 	var err error
-	switch res.Scheme {
-	case NoScheme, "":
-		res.verifier = nil
-	case Ecdsa:
-		res.verifier, err = NewEcdsaVerifier(key)
-	case Bls:
-		res.verifier, err = NewBLSVerifier(key)
-	case Eddsa:
-		res.verifier = &EdDSAVerifier{PublicKey: key}
+
+	switch r := p.GetRule().(type) {
+	case *protoblocktx.NamespacePolicy_ThresholdRule:
+		policy := r.ThresholdRule
+
+		switch policy.Scheme {
+		case NoScheme, "":
+			res.verifier = nil
+		case Ecdsa:
+			res.verifier, err = newEcdsaVerifier(policy.PublicKey)
+		case Bls:
+			res.verifier, err = newBLSVerifier(policy.PublicKey)
+		case Eddsa:
+			res.verifier = &edDSAVerifier{PublicKey: policy.PublicKey}
+		default:
+			return nil, errors.Newf("scheme '%v' not supported", policy.Scheme)
+		}
+	case *protoblocktx.NamespacePolicy_SignatureRule:
+		pp := cauthdsl.NewPolicyProvider(idDeserializer)
+		res.verifier, _, err = pp.NewPolicy(r.SignatureRule)
 	default:
-		return nil, errors.Newf("scheme '%v' not supported", scheme)
+		return nil, errors.Newf("policy rule '%v' not supported", p.GetRule())
 	}
 	return res, err
 }
 
 // VerifyNs verifies a transaction's namespace signature.
 func (v *NsVerifier) VerifyNs(txID string, tx *protoblocktx.Tx, nsIndex int) error {
-	if nsIndex < 0 || nsIndex >= len(tx.Namespaces) || nsIndex >= len(tx.Signatures) {
+	if nsIndex < 0 || nsIndex >= len(tx.Namespaces) || nsIndex >= len(tx.Endorsements) {
 		return errors.New("namespace index out of range")
 	}
+
 	if v.verifier == nil {
 		return nil
 	}
-	digest, err := DigestTxNamespace(txID, tx.Namespaces[nsIndex])
+
+	data, err := ASN1MarshalTxNamespace(txID, tx.Namespaces[nsIndex])
 	if err != nil {
 		return err
 	}
-	return v.verifier.Verify(digest, tx.Signatures[nsIndex])
+
+	switch v.NamespacePolicy.GetRule().(type) {
+	case *protoblocktx.NamespacePolicy_ThresholdRule:
+		return v.verifier.EvaluateSignedData([]*protoutil.SignedData{
+			{
+				Data:      data,
+				Signature: tx.Endorsements[nsIndex].EndorsementsWithIdentity[0].Endorsement,
```

Comment:
major: We should verify that the `EndorsementsWithIdentity` array have exactly one item, and return error otherwise.

Reply:
What if it contains two or more signatures and the first one is the valid one. Should we invalidate the transaction?